crates/subspace-archiving/src/piece_reconstructor.rs

```diff
@@ -0,0 +1,176 @@
+    /// automatically due to use of erasure coding if needed).
+    pub fn retrieve_pieces(
+        &self,
+        segment_pieces: &[Option<Piece>],
+    ) -> Result<Vec<Piece>, PieceReconstructorError> {
+        // If not all data pieces are available, need to reconstruct data shards using erasure
+        // coding.
+        let mut shards = segment_pieces
+            .iter()
+            .map(|maybe_piece| maybe_piece.as_ref().map(utils::slice_to_arrays))
+            .collect::<Vec<_>>();
+
+        self.reed_solomon
+            .reconstruct(&mut shards)
+            .map_err(PieceReconstructorError::DataShardsReconstruction)?;
+
+        let reconstructed_record_shards = shards
+            .iter()
+            .map(|maybe_shard| {
+                maybe_shard
+                    .as_ref()
+                    .map(|shard| {
+                        let bytes = shard.iter().fold(Vec::new(), |mut acc, shard_part| {
+                            acc.extend_from_slice(shard_part);
+
+                            acc
+                        });
+
+                        let record = &bytes[0..self.record_size as usize];
+
+                        record.to_vec()
+                    })
+                    .expect("Record must be reconstructed here.")
+            })
+            .collect::<Vec<_>>();
+
+        let record_shards_hashes = reconstructed_record_shards
+            .iter()
+            .map(|item| blake2b_256_254_hash(item))
+            .collect::<Vec<_>>();
+
+        let data = {
+            let mut data = Vec::with_capacity(
+                (self.data_shards() + self.parity_shards()) as usize * BLAKE2B_256_HASH_SIZE,
+            );
+
+            for shard in &record_shards_hashes {
+                data.extend_from_slice(shard);
+            }
+
+            data
+        };
+        let polynomial = self
+            .kzg
+            .poly(&data)
+            .expect("Internally produced values must never fail; qed");
+
+        let mut pieces = FlatPieces::new(reconstructed_record_shards.len());
+        pieces
+            .as_pieces_mut()
+            .enumerate()
+            .zip(reconstructed_record_shards.iter())
+            .for_each(|((position, piece), shard_chunk)| {
+                let (record_part, witness_part) = piece.split_at_mut(self.record_size as usize);
+
+                record_part.copy_from_slice(shard_chunk);
+                witness_part.copy_from_slice(
+                    &self
+                        .kzg
+                        .create_witness(&polynomial, position as u32)
+                        .expect("We use the same indexes as during Merkle tree creation; qed")
+                        .to_bytes(),
+                );
+            });
+
+        let reconstructed_pieces = pieces
+            .as_pieces()
+            .map(|bytes| Piece::try_from(bytes).expect("Piece must have the correct size"))
+            .collect::<Vec<_>>();
```

Comment:
I'd either return `FlatPieces` here or allocate individual pieces to begin with. The whole point of it is to reduce number of allocations by 256x, but with this we have 257 allocations instead :upside_down_face: